src/lua/mod.rs

```diff
@@ -0,0 +1,283 @@
+//! Lua functionality
+
+use hlua;
+use hlua::{Lua, LuaError};
+use hlua::any::AnyLuaValue;
+
+use std::thread;
+
+use std::fs::{File};
+use std::path::Path;
+use std::io::Write;
+
+use std::sync::{Mutex, RwLock};
+use std::sync::mpsc::{channel, Sender, Receiver};
+
+#[macro_use]
+mod funcs;
+
+lazy_static! {
+    /// Sends requests to the lua thread
+    static ref SENDER: Mutex<Sender<LuaQuery>> = {
+        let (tx, rx) = channel::<LuaQuery>();
+        Mutex::new(tx)
+    };
+
+    /// Receives data back from the lua thread
+    /// This should only be accessed by the lua thread itself.
+    static ref RECEIVER: Mutex<Receiver<LuaResponse>> = {
+        let (tx, rx) = channel::<LuaResponse>();
+        Mutex::new(rx)
+    };
+
+    /// Whether the lua thread is currently running
+    pub static ref RUNNING: RwLock<bool> = RwLock::new(false);
+}
+
+/// Messages sent to the lua thread
+pub enum LuaQuery {
+    /// Halt the lua thread
+    Terminate,
+    // Restart the lua thread
+    Restart,
+    /// Execute a string
+    Execute(String),
+    /// Execute a file
+    ExecuteFile(String),
+    /// Get a variable
+    GetVariable(String),
+    /// Set a value
+    SetValue {
+        name: Box<::std::borrow::Borrow<str> + Sized>,
+        val: Box<hlua::Push<&'static mut Lua<'static>> + Sized>
+    },
+    /// Create a new array
+    EmptyArray(String),
+    /// Message to ping the lua thread
+    Ping,
+    /// Unused send type
+    Unused,
+}
+
+/// Messages received from lua thread
+pub enum LuaResponse {
+    /// Lua variable obtained
+    Variable(Option<AnyLuaValue>),
+    /// Lua error
+    Error(hlua::LuaError),
+    /// A function is returned
+    Function(hlua::functions_read::LuaFunction<String>),
+    /// Pong response from lua ping
+    Pong,
+    /// Unused response type
+    Unused,
+}
+
+unsafe impl Send for LuaQuery { }
+unsafe impl Send for LuaResponse { }
+unsafe impl Sync for LuaQuery { }
+unsafe impl Sync for LuaResponse { }
+
+/// Whether the lua thread is currently available
+pub fn thread_running() -> bool {
+    *RUNNING.read().unwrap()
+}
+
+/// Errors which may arise from attempting
+/// to sending a message to the lua thread.
+pub enum LuaSendError {
+    ThreadClosed,
+    Sender
+}
+
+/// Attemps to send a LuaQuery to the lua thread.
+pub fn try_send(query: LuaQuery) -> Result<(), LuaSendError> {
+    if !thread_running() { Err(LuaSendError::ThreadClosed) }
+    else {
+        match SENDER.lock().unwrap().send(query) {
+            Ok(_) => Ok(()),
+            Err(_) => Err(LuaSendError::Sender)
+        }
+    }
+}
+
+/// Sends a value to the lua thread.
+///
+/// # Panics
+/// * If the lua thread is currently not running (check `lua::thread_running()`)
+/// * If the sender was unable to send, which should only happen
+/// if the lua thread isn't running.
+pub fn send(query: LuaQuery) {
```

Comment:
Do we need to expose this function? `try_send` is already public, if there is a case where some rust code should panic then it can be handled by the caller (by simply unwrapping the result of `try_send`, for example).